File: src/seam-carver.ts
import GridCalculator from "./grid-calculator"

export default class SeamCarver {
  private static BORDER_ENERGY = 1000
  distTo: Array<number>
  edgeTo: Array<number>
  picture: Uint8ClampedArray
  pixelEnergies: Uint16Array
  grid: GridCalculator

  constructor(picture: Uint8ClampedArray, pictureWidth: number, pictureHeight: number) {
    this.grid = new GridCalculator(pictureWidth, pictureHeight)
    this.pixelEnergies = new Uint16Array(this.grid.getLength())
    this.picture = picture
    for(let v = 0; v < this.grid.getLength(); v++) {
      this.pixelEnergies[v] = this.getVertexEnergy(v)
    }
  }

  public width(): number {
    return this.grid.width()
  }

  public height(): number {
    return this.grid.height()
  }

  public getVerticalSeam(): Array<number> {
    this.initDistTo(true);
    this.initEdgeTo();

    for(let v = 0; v < this.grid.getLength(); v++) {
      for (const w of this.grid.adjDownward(v)) this.relax(v, w);
    }

    let minIndex = -1;
    let minValue = Number.POSITIVE_INFINITY;
    const lastRow = this.grid.height() - 1;
    for (let col = 0; col < this.grid.width(); col++) {
      const index = this.grid.getIndex(lastRow, col);
      if (this.distTo[index] < minValue) {
        minValue = this.distTo[index];
        minIndex = index;
      }
    }

    const seam: Array<number> = []
    let seamIter: number = minIndex;
    for (let i = this.grid.height() - 1; i >= 0; i--) {
      seam[i] = this.grid.getColumnOfIndex(seamIter);
      seamIter = this.edgeTo[seamIter];
    }

    return seam;
  }

  public removeVerticalSeam(seam: Array<number>): void {
    const newWidth = this.grid.width() - 1;
    const newPictureColors = new Uint8ClampedArray(newWidth * this.grid.height() * 4);
    let copyOffset = 0, sliceStart = 0, sliceEnd = 0
    for (let row = 0; row < seam.length; row++) {
      sliceEnd = this.grid.getIndex(row, seam[row]) * 4
      newPictureColors.set(this.picture.slice(sliceStart, sliceEnd), copyOffset)
      copyOffset += (sliceEnd - sliceStart)
      sliceStart = sliceEnd + 4
    }
    this.picture = newPictureColors;
    this.removePixelEnergySeam(seam, newWidth)
    this.grid = new GridCalculator(newWidth, this.grid.height())
  }

  private removePixelEnergySeam(seam: Array<number>, newWidth: number) {
    const newPixelEnergies = new Uint16Array(newWidth * this.grid.height())
    let copyOffset = 0, sliceStart = 0, sliceEnd = 0
    for(let row = 0; row < seam.length; row++) {
      sliceEnd = this.grid.getIndex(row, seam[row])
      newPixelEnergies.set(this.pixelEnergies.slice(sliceStart, sliceEnd), copyOffset)
      copyOffset += (sliceEnd - sliceStart)
      sliceStart = sliceEnd + 1 
    }
    this.pixelEnergies = newPixelEnergies
  }

  private initDistTo(verticalOrientation: boolean) {
    this.distTo = []
    for (let i = 0; i < this.grid.getLength(); i++) {
      this.distTo[i] = Number.MAX_VALUE
      if (verticalOrientation) { // vertical orientation => first row dist = 0
        if (i < this.grid.width()) {
          this.distTo[i] = 0;
        }
      } else if (i % this.grid.width() == 0) { // horizontal orientation => first col dist = 0
        this.distTo[i] = 0;
      }
    }
  }

  private initEdgeTo() {
    this.edgeTo = []
    for (let i = 0; i < this.grid.getLength(); i++) {
      this.edgeTo[i] = -1;
    }
  }

  private relax(v: number, w: number) {
    const wEnergy = this.pixelEnergies[w];
    if (this.distTo[w] > this.distTo[v] + wEnergy) {
      this.distTo[w] = this.distTo[v] + wEnergy;
      this.edgeTo[w] = v;
    }
  }

  public energy(col: number, row: number): number {
    if (row == 0 || row == this.grid.height() - 1 || col == 0 || col == this.grid.width() - 1) {
      return SeamCarver.BORDER_ENERGY
    }

    const deltaRow = this.getColorDelta(this.getColor(col, row + 1), this.getColor(col, row - 1))
    const deltaCol = this.getColorDelta(this.getColor(col + 1, row), this.getColor(col - 1, row))

    return Math.sqrt(deltaRow + deltaCol);
  }

  private getVertexEnergy(vertex: number) {
    return this.energy(this.grid.getColumnOfIndex(vertex), this.grid.getRowOfIndex(vertex))
  }

  private getColorDelta(rgbaA: Uint8ClampedArray, rgbaB: Uint8ClampedArray) {
    return (
      Math.pow((rgbaA[0]) - (rgbaB[0]), 2) +
      Math.pow((rgbaA[1]) - (rgbaB[1]), 2) +
      Math.pow((rgbaA[2]) - (rgbaB[2]), 2)
    )
  }

  private getColor(col: number, row: number): Uint8ClampedArray {
    const redIndex = this.grid.getIndex(row, col) * 4
    return this.picture.slice(redIndex, redIndex + 4)
  }
}

File: golang/seam-carver.go
package main

import (
	"math"
)

const BORDER_ENERGY = 1000

type Color = [4]uint8

type SeamCarver struct {
	distTo        []int
	edgeTo        []int
	picture       []uint8
	pixelEnergies []uint16
	grid          GridCalculator
}

func NewSeamCarver(picture []uint8, pictureWidth int, pictureHeight int) SeamCarver {
	seamCarver := SeamCarver{
		picture: picture,
		grid:    GridCalculator{width: pictureWidth, height: pictureHeight},
	}
	seamCarver.initPixelEnergies()

	return seamCarver
}

func (sc SeamCarver) GetVerticalSeam() []uint16 {
	sc.initDistTo(true)
	sc.initEdgeTo()

	length := sc.grid.GetLength()
	for v := 0; v < length; v++ {
		sc.relaxNeighbors(v)
	}

	minIndex := -1
	minValue := math.MaxInt
	lastRow := sc.grid.height - 1
	for col := 0; col < sc.grid.width; col++ {
		index := sc.grid.GetIndex(lastRow, col)
		if sc.distTo[index] < minValue {
			minValue = sc.distTo[index]
			minIndex = index
		}
	}

	seam := make([]uint16, sc.grid.height)
	seamIter := minIndex
	for i := sc.grid.height - 1; i >= 0; i-- {
		seam[i] = uint16(sc.grid.GetColumnFromIndex(seamIter))
		seamIter = sc.edgeTo[seamIter]
	}

	return seam
}

func (sc *SeamCarver) RemoveVerticalSeam(seam []uint16) {
	newWidth := sc.grid.width - 1
	newPictureColors := make([]uint8, newWidth*sc.grid.height*4)
	copyOffset := 0
	sliceStart := 0
	sliceEnd := 0
	for row := 0; row < len(seam); row++ {
		sliceEnd = sc.grid.GetIndex(row, int(seam[row])) * 4
		copy(newPictureColors[copyOffset:], sc.picture[sliceStart:sliceEnd])
		copyOffset += (sliceEnd - sliceStart)
		sliceStart = sliceEnd + 4
	}
	sc.picture = newPictureColors
	sc.removePixelEnergySeam(seam, newWidth)
	sc.grid = GridCalculator{width: newWidth, height: sc.grid.height}
}

func (sc *SeamCarver) initPixelEnergies() {
	length := sc.grid.GetLength()
	sc.pixelEnergies = make([]uint16, length)

	for v := 0; v < length; v++ {
		sc.pixelEnergies[v] = sc.getVertexEnergy(sc.grid.GetColumnFromIndex(v), sc.grid.GetRowFromIndex(v))
	}
}

func (sc SeamCarver) getVertexEnergy(col int, row int) uint16 {
	if row == 0 || row == sc.grid.height-1 || col == 0 || col == sc.grid.width-1 {
		return BORDER_ENERGY
	}

	deltaRow := sc.getColorDelta(sc.getColor(col, row+1), sc.getColor(col, row-1))
	deltaCol := sc.getColorDelta(sc.getColor(col+1, row), sc.getColor(col-1, row))

	// TODO add check conversion is not overflowing uint16
	return uint16(math.Sqrt(deltaRow + deltaCol))
}

func (sc SeamCarver) getColor(col int, row int) Color {
	redIndex := sc.grid.GetIndex(row, col) * 4
	return Color{
		sc.picture[redIndex],
		sc.picture[redIndex+1],
		sc.picture[redIndex+2],
		sc.picture[redIndex+3],
	}
}

func (sc SeamCarver) getColorDelta(a Color, b Color) float64 {
	// TODO add alpha diff?
	return math.Pow(float64(subsUint8(a[0], b[0])), 2) +
		math.Pow(float64(subsUint8(a[1], b[1])), 2) +
		math.Pow(float64(subsUint8(a[2], b[2])), 2)
}

func subsUint8(a uint8, b uint8) uint8 {
	if a > b {
		return a - b
	} else {
		return b - a
	}
}

func (sc *SeamCarver) removePixelEnergySeam(seam []uint16, newWidth int) {
	newPixelEnergies := make([]uint16, newWidth*sc.grid.height)
	copyOffset := 0
	sliceStart := 0
	sliceEnd := 0
	for row := 0; row < len(seam); row++ {
		sliceEnd = sc.grid.GetIndex(row, int(seam[row]))
		copy(newPixelEnergies[copyOffset:], sc.pixelEnergies[sliceStart:sliceEnd])
		copyOffset += (sliceEnd - sliceStart)
		sliceStart = sliceEnd + 1
	}
	sc.pixelEnergies = newPixelEnergies
}

func (sc *SeamCarver) initDistTo(verticalOrientation bool) {
	length := sc.grid.GetLength()
	sc.distTo = make([]int, length)
	for i := 0; i < sc.grid.GetLength(); i++ {
		sc.distTo[i] = math.MaxInt
		if verticalOrientation { // vertical orientation => first row dist = 0
			if i < sc.grid.width {
				sc.distTo[i] = 0
			}
		} else if i%sc.grid.width == 0 { // horizontal orientation => first col dist = 0
			sc.distTo[i] = 0
		}
	}
}

func (sc *SeamCarver) initEdgeTo() {
	length := sc.grid.GetLength()
	sc.edgeTo = make([]int, length)
	for i := 0; i < length; i++ {
		sc.edgeTo[i] = -1
	}
}

func (sc *SeamCarver) relaxNeighbors(v int) {
	row := sc.grid.GetRowFromIndex(v)
	col := sc.grid.GetColumnFromIndex(v)

	if row >= sc.grid.height-1 {
		return
	}

	// bottom neighbor
	sc.relax(v, sc.grid.GetIndex(row+1, col))

	// bottom-left neighbor
	if col > 0 {
		sc.relax(v, sc.grid.GetIndex(row+1, col-1))
	}
	// bottom-right neighbor.
	if col < sc.grid.width-1 {
		sc.relax(v, sc.grid.GetIndex(row+1, col+1))
	}
}

func (sc *SeamCarver) relax(v int, w int) {
	wEnergy := int(sc.pixelEnergies[w])
	if sc.distTo[w] > sc.distTo[v]+wEnergy {
		sc.distTo[w] = sc.distTo[v] + wEnergy
		sc.edgeTo[w] = v
	}
}
